src/gt4py/next/iterator/transforms/global_tmps.py

```diff
@@ -373,103 +348,179 @@ def _max_domain_sizes_by_location_type(offset_provider: Mapping[str, Any]) -> di
+            assert isinstance(axis_literal, ir.AxisLiteral)
+
+            ranges[axis_literal.value] = SymbolicRange(lower_bound, upper_bound)
+        return cls(node.fun.id, ranges)  # type: ignore[attr-defined]  # ensure by assert above
+
+    def as_expr(self):
+        return im.call(self.grid_type)(
+            *[
+                im.call("named_range")(ir.AxisLiteral(value=d), r.start, r.stop)
+                for d, r in self.ranges.items()
+            ]
+        )
+
+
+def update_domains(node: FencilWithTemporaries, offset_provider: Mapping[str, Any]):
     horizontal_sizes = _max_domain_sizes_by_location_type(offset_provider)
-    vertical_ranges = _domain_ranges(node.fencil.closures)
-    for k in horizontal_sizes:
-        vertical_ranges.pop(k, None)
 
-    closures = []
-    domains = dict[str, ir.Expr]()
+    closures: list[ir.StencilClosure] = []
+    domains = dict[str, ir.FunCall]()
     for closure in reversed(node.fencil.closures):
         if closure.domain == AUTO_DOMAIN:
+            # every closure with auto domain should have a single out field
+            assert isinstance(closure.output, ir.SymRef)
             domain = domains[closure.output.id]
             closure = ir.StencilClosure(
-                domain=domain, stencil=closure.stencil, output=closure.output, inputs=closure.inputs
+                domain=copy.deepcopy(domain),
+                stencil=closure.stencil,
+                output=closure.output,
+                inputs=closure.inputs,
             )
         else:
             domain = closure.domain
 
         closures.append(closure)
 
         if closure.stencil == ir.SymRef(id="deref"):
-            domains[closure.inputs[0].id] = domain
+            # all closure inputs inherit the domain
+            for input_arg in _tuple_constituents(closure.inputs[0]):
+                assert isinstance(input_arg, ir.SymRef)
+                assert domains.get(input_arg.id, domain) == domain
+                domains[input_arg.id] = domain
             continue
 
         local_shifts = TraceShifts.apply(closure)
-        for param, shifts in local_shifts.items():
-            loctypes = {_location_type_from_offsets(domain, s, offset_provider) for s in shifts}
-            assert len(loctypes) == 1
-            loctype = loctypes.pop()
-            horizontal_size = horizontal_sizes[loctype]
-            domains[param] = _unstructured_domain(
-                loctype, horizontal_size, vertical_ranges.values()
+        for param, shift_chains in local_shifts.items():
+            assert isinstance(param, str)
+            consumed_domains: list[SymbolicDomain] = (
+                [SymbolicDomain.from_expr(domains[param])] if param in domains else []
             )
+            for shift_chain in shift_chains:
+                consumed_domain = SymbolicDomain.from_expr(domain)
+                for shift in zip(shift_chain[::2], shift_chain[1::2], strict=True):
```

Comment:
this is hard to read if you don't know the internals very well -> `group_offsets(offsets: Sequence[Offset]) -> tuple[tag,int]`